pkg/linters/components/compimpl/linter.go

```diff
@@ -0,0 +1,164 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// Package compimpl is a linter that checks if the component implementation follows the expected naming convention.
+package compimpl
+
+import (
+	"fmt"
+	"go/ast"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"golang.org/x/tools/go/analysis"
+)
+
+func init() {
+	RegisterLinter("compimpl", "Check the component implementation follows the naming convention", run)
+}
+
+type implFolder struct {
+	hasProvides    bool
+	HasRequires    bool
+	HasConstructor bool
+	file           *ast.File // The first file found in the folder. Used to display the diagnostic
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	// The regexp to match path like comp/core/autodiscovery/noopimpl/autoconfig.go but not the folders inside noopimpl
+	isCompRegexp := regexp.MustCompile("(.*/comp/.*/.*/.*impl)/[^/]*$")
+	compimpls := make(map[string]*implFolder)
+
+	for _, f := range pass.Files {
+		implFolder := tryGetCompImpl(pass, f, isCompRegexp, compimpls)
+		if implFolder == nil {
+			continue
+		}
+
+		structTypeNames := getExportedStructTypeNames(f)
+		implFolder.hasProvides = implFolder.hasProvides || structTypeNames["Provides"]
+		implFolder.HasRequires = implFolder.HasRequires || structTypeNames["Requires"]
+		implFolder.HasConstructor = implFolder.HasConstructor || hasComponentConstructor(f)
+	}
+	displayDiagnostic(pass, compimpls)
+	return nil, nil
+}
+
+func displayDiagnostic(pass *analysis.Pass, compimpls map[string]*implFolder) {
+	for folderName, implFolder := range compimpls {
+		file := implFolder.file
+		if !implFolder.hasProvides {
+			pass.Report(analysis.Diagnostic{
+				Pos:     file.Pos(),
+				Message: missingProvides(folderName),
+			})
+		}
+		if !implFolder.HasRequires {
+			pass.Report(analysis.Diagnostic{
+				Pos:     file.Pos(),
+				Message: missingRequires(folderName),
+			})
+		}
+		if !implFolder.HasConstructor {
+			pass.Report(analysis.Diagnostic{
+				Pos:     file.Pos(),
+				Message: missingConstructor(folderName),
+			})
+		}
+	}
+}
+
+func tryGetCompImpl(
+	pass *analysis.Pass,
+	file *ast.File,
+	isCompRegexp *regexp.Regexp,
+	compimpls map[string]*implFolder) *implFolder {
+	path := pass.Fset.Position(file.Pos()).Filename
+	capture := isCompRegexp.FindStringSubmatch(path)
+	if len(capture) != 2 {
+		return nil
+	}
```

Comment:
We could probably move the the check of the regex outside of this function that way we do not have to pass the regex variable 😄